internal/ls/codelens.go

```diff
@@ -0,0 +1,218 @@
+	textDoc := lsproto.TextDocumentIdentifier{
+		Uri: uri,
+	}
+	locale := locale.FromContext(ctx)
+
+	var locs []lsproto.Location
+	var lensTitle string
+	switch codeLens.Data.Kind {
+	case lsproto.CodeLensKindReferences:
+		origNode, symbolsAndEntries, ok := l.ProvideSymbolsAndEntries(ctx, uri, codeLens.Range.Start, false /*isRename*/)
+		if ok {
+			references, err := l.ProvideReferencesFromSymbolAndEntries(
+				ctx,
+				&lsproto.ReferenceParams{
+					TextDocument: textDoc,
+					Position:     codeLens.Range.Start,
+					Context: &lsproto.ReferenceContext{
+						// Don't include the declaration in the references count.
+						IncludeDeclaration: false,
+					},
+				},
+				origNode,
+				symbolsAndEntries,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			if references.Locations != nil {
+				locs = *references.Locations
+			}
+		}
+
+		if len(locs) == 1 {
+			lensTitle = diagnostics.X_1_reference.Localize(locale)
+		} else {
+			lensTitle = diagnostics.X_0_references.Localize(locale, len(locs))
+		}
+	case lsproto.CodeLensKindImplementations:
+		// "Force" link support to be false so that we only get `Locations` back,
+		// and don't include the "current" node in the results.
+		findImplsOptions := provideImplementationsOpts{
+			requireLocationsResult: true,
+			dropOriginNodes:        true,
+		}
+		implementations, err := l.provideImplementationsEx(
+			ctx,
+			&lsproto.ImplementationParams{
+				TextDocument: textDoc,
+				Position:     codeLens.Range.Start,
+			},
+			findImplsOptions,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if implementations.Locations != nil {
+			locs = *implementations.Locations
+		}
+
+		if len(locs) == 1 {
+			lensTitle = diagnostics.X_1_implementation.Localize(locale)
+		} else {
+			lensTitle = diagnostics.X_0_implementations.Localize(locale, len(locs))
+		}
+	}
+
+	cmd := &lsproto.Command{
+		Title: lensTitle,
+	}
+	if len(locs) > 0 && showLocationsCommandName != nil {
+		cmd.Command = *showLocationsCommandName
+		cmd.Arguments = &[]any{
+			uri,
+			codeLens.Range.Start,
+			locs,
+		}
+	}
+
+	codeLens.Command = cmd
+	return codeLens, nil
+}
+
+func (l *LanguageService) newCodeLensForNode(fileUri lsproto.DocumentUri, file *ast.SourceFile, node *ast.Node, kind lsproto.CodeLensKind) *lsproto.CodeLens {
+	nodeForRange := node
+	nodeName := node.Name()
+	if nodeName != nil {
+		nodeForRange = nodeName
+	}
+	pos := scanner.SkipTrivia(file.Text(), nodeForRange.Pos())
+
+	return &lsproto.CodeLens{
+		Range: lsproto.Range{
+			Start: l.converters.PositionToLineAndCharacter(file, core.TextPos(pos)),
+			End:   l.converters.PositionToLineAndCharacter(file, core.TextPos(node.End())),
+		},
+		Data: &lsproto.CodeLensData{
+			Kind: kind,
+			Uri:  fileUri,
+		},
+	}
+}
+
+func isValidImplementationsCodeLensNode(node *ast.Node, userPrefs *lsutil.UserPreferences) bool {
+	switch node.Kind {
+	// Always show on interfaces
+	case ast.KindInterfaceDeclaration:
+		// TODO: ast.KindTypeAliasDeclaration?
```

Comment:
Probably won't address this here. VS Code just didn't show them on type aliases.